### ISSUE TITLE
Panic when sleeping until far into the future 

### ISSUE THREAD

**Issue by Erk-** [tokio-rs/tokio#5183]
**Version**

At least `1.21.2` and current git `master`

**Platform**

`Linux archlinux 6.0.7-arch1-1 #1 SMP PREEMPT_DYNAMIC Thu, 03 Nov 2022 18:01:58 +0000 x86_64 GNU/Linux`

**Description**

Panic when sleeping until far into the future (though not if it is too far)

I tried this code:

```rust
#[tokio::test]
async fn t() {
    // u64::MAX will work as expected.
    let big = std::time::Duration::from_secs(u64::MAX/10);
    tokio::time::sleep(big).await;
}
```
[Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=83b601b77fdab8441087721b0b6fe303)

I expected to see this happen: Nothing, in that it would just sleep

Instead, this happened:
In debug it hits a debug assertion:
```
thread 't' panicked at 'assertion failed: timestamp < STATE_MIN_VALUE', /home/valdemar/.cargo/git/checkouts/tokio-377c595163f99a10/9e3fb16/tokio/src/runtime/time/entry.rs:239:9
```

In release mode it hits another panic:
```
thread 't' panicked at 'Timer already fired', /home/valdemar/.cargo/git/checkouts/tokio-377c595163f99a10/9e3fb16/tokio/src/runtime/time/entry.rs:398:9
```

**Extra context**

This seems to be similar to the issue solved by https://github.com/tokio-rs/tokio/pull/4495.

But it seems that the use of `u64::MAX` causes issues because it is used as signal values.

The issue seems to be resolve by changing the value [here](https://github.com/tokio-rs/tokio/blob/9e3fb1673a73046363af64c09a040eeed67f2a4c/tokio/src/runtime/time/source.rs#L32) into `u64::MAX - 1`, but I am not sure if that is the best way to resolve this issue, given it also seem to be used as a signal value [here](https://github.com/tokio-rs/tokio/blob/tokio-1.21.x/tokio/src/time/driver/entry.rs#L74).

**Comment by Simon-Laux** [comment]
> The issue seems to be resolve by changing the value [here](https://github.com/tokio-rs/tokio/blob/9e3fb1673a73046363af64c09a040eeed67f2a4c/tokio/src/runtime/time/source.rs#L32) into u64::MAX - 1, but I am not sure if that is the best way to resolve this issue, given it also seem to be used as a signal value [here](https://github.com/tokio-rs/tokio/blob/tokio-1.21.x/tokio/src/time/driver/entry.rs#L74).

Could `u64::MAX - 2` work then? I think it would make sense to define whatever is used as maximum in the end as constant.

**Comment by Erk-** [comment]
> > The issue seems to be resolve by changing the value [here](https://github.com/tokio-rs/tokio/blob/9e3fb1673a73046363af64c09a040eeed67f2a4c/tokio/src/runtime/time/source.rs#L32) into u64::MAX - 1, but I am not sure if that is the best way to resolve this issue, given it also seem to be used as a signal value [here](https://github.com/tokio-rs/tokio/blob/tokio-1.21.x/tokio/src/time/driver/entry.rs#L74).
> 
> Could `u64::MAX - 2` work then? I think it would make sense to define whatever is used as maximum in the end as constant.

Yeah that is how I have solved it, but I had not made a pr because the test I have added blocks indefinitely.

I have opened a PR with a solution to the bug, but with the broken test in #5710.

**Comment by Erk-** [comment]
This was fixed in #5710
